src/lib/zexe_backend/zexe_backend_common/plonk_plookup_constraint_system.ml

```diff
@@ -0,0 +1,1172 @@
+               ; Fp.zero
+               ; Fp.zero
+               ; Fp.zero
+               ; Fp.zero
+               ; Fp.negate s3 |]
+              sys
+        | `Constant, `Constant, `Var x3 ->
+            add_generic_constraint ~o:x3
+              [| Fp.zero
+               ; Fp.zero
+               ; s3
+               ; Fp.zero
+               ; Fp.zero
+               ; Fp.zero
+               ; Fp.(negate s1 * s2) |]
+              sys
+        | `Constant, `Constant, `Constant ->
+            assert (Fp.(equal s3 Fp.(s1 * s2))) )
+    | Snarky_backendless.Constraint.Boolean v -> (
+        let s, x = red v in
+        match x with
+        | `Var x ->
+            (* -x + x * x = 0  *)
+            add_generic_constraint ~l:x ~r:x
+              [| Fp.(negate one)
+               ; Fp.zero
+               ; Fp.zero
+               ; Fp.zero
+               ; Fp.zero
+               ; Fp.one
+               ; Fp.zero |]
+              sys
+        | `Constant ->
+            assert (Fp.(equal s (s * s))) )
+    | Snarky_backendless.Constraint.Equal (v1, v2) -> (
+        let (s1, x1), (s2, x2) = (red v1, red v2) in
+        match (x1, x2) with
+        | `Var x1, `Var x2 ->
+            (* s1 x1 - s2 x2 = 0
+          *)
+            if s1 <> s2 then
+              add_generic_constraint ~l:x1 ~r:x2
+                [| s1
+                 ; Fp.(negate s2)
+                 ; Fp.zero
+                 ; Fp.zero
+                 ; Fp.zero
+                 ; Fp.zero
+                 ; Fp.zero |]
+                sys
+              (* TODO: optimize by not adding generic costraint but rather permuting the vars *)
+            else
+              add_generic_constraint ~l:x1 ~r:x2
+                [| s1
+                 ; Fp.(negate s2)
+                 ; Fp.zero
+                 ; Fp.zero
+                 ; Fp.zero
+                 ; Fp.zero
+                 ; Fp.zero |]
+                sys
+        | `Var x1, `Constant ->
+            add_generic_constraint ~l:x1
+              [|s1; Fp.zero; Fp.zero; Fp.zero; Fp.zero; Fp.zero; Fp.negate s2|]
+              sys
+        | `Constant, `Var x2 ->
+            add_generic_constraint ~r:x2
+              [|Fp.zero; s2; Fp.zero; Fp.zero; Fp.zero; Fp.zero; Fp.negate s1|]
+              sys
+        | `Constant, `Constant ->
+            assert (Fp.(equal s1 s2)) )
+    | Plonk_constraint.T (Basic {l; r; o; m; c}) ->
+        (* 0
+         = l.s * l.x
+         + r.s * r.x
+         + o.s * o.x
+         + m * (l.x * r.x)
+         + c
+         =
+           l.s * l.s' * l.x'
+         + r.s * r.s' * r.x'
+         + o.s * o.s' * o.x'
+         + m * (l.s' * l.x' * r.s' * r.x')
+         + c
+         =
+           (l.s * l.s') * l.x'
+         + (r.s * r.s') * r.x'
+         + (o.s * o.s') * o.x'
+         + (m * l.s' * r.s') * l.x' r.x'
+         + c
+      *)
+        (* TODO: This is sub-optimal *)
+        let c = ref c in
+        let red_pr (s, x) =
```

Comment:
This could do with a clearer name (although I suspect this is true of the current plonk CS too, so perhaps here is not the place to fix it).